From 189ea653ff1da0b7bb8fa5a13deab360a285eddd Mon Sep 17 00:00:00 2001
From: Jiffin Tony Thottan <jthottan@redhat.com>
Date: Wed, 5 Aug 2015 16:56:46 +0530
Subject: [PATCH] Moving up Clean_RPC() to the top of do_shutdown().

As mentioned in the mail thread http://sourceforge.net/p/nfs-ganesha/
mailman/message/34345947/, there is high chance of ganesha process got
killed before completing all the steps mentioned in the do_shutdown().
So if ganesha got killed before cleaning up the ports,other nfs-services
cannot come up unless user clean up those ports manually.

Normally ganesha got killed while waiting up for worker threads to close.
Therefore it is better to clean up the ports before closing all threads.
So that atleast we can make sure that ganesha is no longer using ports
which might needed by other application to run.

Change-Id: I97723f6c0275f8d8feac0ca7ea87d2509c5b4b6a
Signed-off-by: Jiffin Tony Thottan <jthottan@redhat.com>
---
 src/MainNFSD/nfs_admin_thread.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/src/MainNFSD/nfs_admin_thread.c b/src/MainNFSD/nfs_admin_thread.c
index 4c82a55..1b60b5a 100644
--- a/src/MainNFSD/nfs_admin_thread.c
+++ b/src/MainNFSD/nfs_admin_thread.c
@@ -334,6 +334,10 @@ static void do_shutdown(void)
 	LogEvent(COMPONENT_MAIN, "Stopping request listener threads.");
 	nfs_rpc_dispatch_stop();
 
+	LogEvent(COMPONENT_MAIN, "Unregistering ports used by NFS service");
+	/* finalize RPC package */
+	Clean_RPC();
+
 	LogEvent(COMPONENT_MAIN, "Stopping request decoder threads");
 	rc = fridgethr_sync_command(req_fridge, fridgethr_comm_stop, 120);
 
@@ -364,8 +368,6 @@ static void do_shutdown(void)
 			 "Worker threads successfully shut down.");
 	}
 
-	/* finalize RPC package */
-	Clean_RPC(); /* we MUST do this first */
 	(void)svc_shutdown(SVC_SHUTDOWN_FLAG_NONE);
 
 	rc = general_fridge_shutdown();
-- 
2.1.0

